### PR TITLE
VicReg Loss - Improvement of Barlow Twins

### PR DESCRIPTION
**Pull request by dewball345**
This is an implementation of the loss of VICReg: Variance-Invariance-Covariance Regularization for Self-Supervised Learning, created by Adrien Bardes, Jean Ponce, and Yann LeCun over at Meta Research. 

This loss function aims to prevent the presence of redundancy and collapse during contrastive learning(much like Barlow Twins). It does this by creating three separate losses and merging them together.

These losses include:

- Invariance Loss: To check if representations are same for same image
- Covariance Loss: Make sure there is no redundancy(Shared with barlow twins)
- Variance Loss: Make sure embeddings contain most amount of information(usually equates to the values having high variance)

### CODE REVIEW

**Comment by dewball345** [comment]
Ill also follow up by making an augmentation pipeline for VicReg(if it is needed though, will have to check paper to see if there are any parameter changes) 

Also: Pseudocode on paper:
https://pbs.twimg.com/media/E1MkDLGXEAUwYuX?format=jpg&name=900x900

**Comment by dewball345** [comment]
Also did some benchmarking(only on CIFAR-10 though) and noted slightly better results. 

With a linear evaluation of Barlow Twins noticed that the loss was around a 66% 
But with VicReg loss is at 67%

I'll do some formal testing later, but for now I can tell that VicReg is promising.

File: tensorflow_similarity/losses/vicreg.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        batch_size = tf.shape(za)[0]
+         
+        # distance loss to measure similarity between representations
+        sim_loss = tf.keras.losses.MeanSquaredError()(za, zb)
+        
+        za = self.standardize_columns(za)
+        zb = self.standardize_columns(zb)
+        
+        # std loss to maximize variance(information)
+        std_za = tf.sqrt(tf.math.reduce_variance(za, 0) + self.std_const)
```
Comment (owenvallis) [inline tensorflow_similarity/losses/vicreg.py:45]: where is self.std_const defined? This seems like the first time it's called.

File: tensorflow_similarity/losses/vicreg.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        return {**base_config, **config}
+
+    def off_diagonal(self, x: FloatTensor) -> FloatTensor:
+        n = tf.shape(x)[0]
+        flattened = tf.reshape(x, [-1])[:-1]
+        off_diagonals = tf.reshape(flattened, (n - 1, n + 1))[:, 1:]
+        off_diag: FloatTensor = tf.reshape(off_diagonals, [-1])
+        return off_diag
+
+    def standardize_columns(self, x: FloatTensor) -> FloatTensor:
```
[the remark this answers is not in the document]
Comment (dewball345) [reply tensorflow_similarity/losses/vicreg.py:96]: Oops, yes, i forgot to change the function name. But you are right, this is just mean-centering

File: tensorflow_similarity/losses/vicreg.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+from tensorflow_similarity.types import FloatTensor
+
+
+@tf.keras.utils.register_keras_serializable(package="Similarity")
+class VicReg(Loss):
+    """VicReg Loss"""
+
+    def __init__(self,
+                 const_std: float = 1e-4,
```
[the remark this answers is not in the document]
Comment (dewball345) [reply tensorflow_similarity/losses/vicreg.py:14]: My bad, should change that.

**Comment by dewball345** [comment]
I will address these comments. Also, sorry about the master branch thing. will move that to development now.

File: tensorflow_similarity/losses/vicreg.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            za: Embedding A
+            zb: Embedding B
+        Returns:
+            loss
+        """
+        # compute the diagonal
+        batch_size = tf.shape(za)[0]
+         
+        # distance loss to measure similarity between representations
+        sim_loss = tf.keras.losses.MeanSquaredError()(za, zb)
```
[the remark this answers is not in the document]
Comment (dewball345) [reply tensorflow_similarity/losses/vicreg.py:39]: I don't think i noticed any difference, but will try that out

**Comment by dewball345** [comment]
Did fixes

**Review by owenvallis — APPROVED** [review]
